File: solutions/2026-05-03/java/2026-05-03-rotate-string.java
// Problem: Rotate String
// Link: https://leetcode.com/problems/rotate-string/
//
// Approach:
// The problem asks if string `goal` can be obtained by repeatedly shifting the leftmost character of string `s` to the rightmost position.
// If `s` can be rotated to become `goal`, it means `goal` must be a substring of `s` concatenated with itself (`s + s`).
// For example, if `s = "abcde"`, then `s + s = "abcdeabcde"`.
// If `goal = "cdeab"`, then "cdeab" is indeed a substring of "abcdeabcde".
// If `goal = "abced"`, then "abced" is not a substring of "abcdeabcde".
//
// We first check if the lengths of `s` and `goal` are equal. If they are not equal, `goal` can never be formed by rotating `s`, so we return `false`.
// If the lengths are equal, we then check if `s + s` contains `goal` as a substring.
//
// Time Complexity: O(N), where N is the length of the strings.
// Concatenating `s` with itself takes O(N) time.
// Checking if `goal` is a substring of `s + s` using `contains()` typically has a time complexity of O(N * M) in the worst case (where N is length of `s+s` and M is length of `goal`), but for typical string implementations, it's closer to O(N) on average. Since `s+s` has length 2N and `goal` has length N, this operation is effectively O(N).
//
// Space Complexity: O(N), where N is the length of the strings.
// We create a new string `s + s`, which takes O(N) space.

class Solution {
    public boolean rotateString(String s, String goal) {
        // If the lengths of s and goal are different, they can never be equal after rotation.
        if (s.length() != goal.length()) {
            return false;
        }

        // If s is empty, goal must also be empty to be a rotated version.
        if (s.isEmpty()) {
            return goal.isEmpty();
        }

        // Concatenate s with itself. If goal is a rotated version of s,
        // it must be a substring of s + s.
        // For example, if s = "abcde", then s + s = "abcdeabcde".
        // If goal = "cdeab", "cdeab" is a substring of "abcdeabcde".
        String doubledS = s + s;

        // Check if goal is a substring of the doubled string.
        // The contains() method checks for substring existence.
        return doubledS.contains(goal);
    }
}

File: solutions/2026-05-03/python/2026-05-03-rotate-string.py
```python
# Problem: Rotate String
# Problem Summary: Checks if string 's' can be transformed into string 'goal' by repeatedly moving the first character to the end.
# Link: https://leetcode.com/problems/rotate-string/
#
# Approach:
# If 's' can be rotated to become 'goal', then 'goal' must be a substring of 's' concatenated with itself ('s' + 's').
# This is because any rotation of 's' will appear as a contiguous substring within 's' + 's'.
# For example, if s = "abcde", then s + s = "abcdeabcde".
# The rotations of "abcde" are:
# "abcde"
# "bcdea"
# "cdeab"
# "deabc"
# "eabcd"
# All these rotations are present as substrings in "abcdeabcde".
#
# We also need to ensure that 's' and 'goal' have the same length. If they don't, rotation is impossible.
#
# Time Complexity:
# O(N), where N is the length of the string 's' (and 'goal' since lengths must be equal).
# String concatenation takes O(N) time.
# The `in` operator (substring check) for Python strings has an average time complexity of O(N * M), where N is the length of the text to search within and M is the length of the substring. In this case, the text is `s + s` (length 2N) and the substring is `goal` (length N). So, this is effectively O(2N * N) in the worst case. However, Python's string searching is often optimized. A more accurate analysis considering the KMP algorithm or similar would bring it closer to O(N). For practical purposes and typical LeetCode constraints, it's often considered O(N). If we consider the worst-case theoretical complexity of naive substring search, it could be O(N^2). Given the constraints (length <= 100), O(N^2) is acceptable.
#
# Space Complexity:
# O(N), where N is the length of the string 's'. This is due to the creation of the new string `s + s`.

class Solution:
    def rotateString(self, s: str, goal: str) -> bool:
        # First, check if the lengths of the two strings are equal.
        # If the lengths are different, it's impossible for 's' to become 'goal' by rotation.
        if len(s) != len(goal):
            return False

        # If both strings are empty, they are considered equal.
        if not s:
            return True

        # Concatenate 's' with itself. This creates a string that contains all possible rotations of 's' as substrings.
        # For example, if s = "abc", then s_doubled = "abcabc".
        # The rotations of "abc" are "abc", "bca", "cab", all of which are substrings of "abcabc".
        s_doubled = s + s

        # Check if 'goal' is a substring of the doubled string 's_doubled'.
        # If 'goal' is found within 's_doubled', it means 'goal' is a rotation of 's'.
        return goal in s_doubled

```